File: backend/app/api/gis.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.database import get_db
from app.gis import service
from app.models.groundwater import AssessmentUnit, District, State, Village
from app.models.user import User

router = APIRouter(prefix="/gis", tags=["gis"])


@router.get("/search")
def gis_search(
    q: str = Query(default="", min_length=1, max_length=120),
    limit: int = Query(default=10, ge=1, le=25),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    """Search states, districts, villages and assessment units by name.

    Returns the top matches with their scope names and coordinates so the map
    can zoom to a searched place.
    """
    needle = f"%{q.strip().lower()}%"
    results: list[dict] = []

    if len(q.strip()) >= 2:
        states = db.execute(
            select(State).where(func.lower(State.name).like(needle)).order_by(State.name).limit(limit)
        ).scalars().all()
        for s in states:
            results.append(
                {
                    "type": "state",
                    "name": s.name,
                    "state": s.name,
                    "district": None,
                    "village": None,
                    "latitude": None,
                    "longitude": None,
                }
            )

        districts = db.execute(
            select(District)
            .join(State, District.state_id == State.id)
            .where(func.lower(District.name).like(needle))
            .order_by(State.name, District.name)
            .limit(limit)
        ).scalars().all()
        for d in districts:
            results.append(
                {
                    "type": "district",
                    "name": d.name,
                    "state": d.state.name,
                    "district": d.name,
                    "village": None,
                    "latitude": None,
                    "longitude": None,
                }
            )

        villages = db.execute(
            select(Village, District, State)
            .join(District, Village.district_id == District.id)
            .join(State, District.state_id == State.id)
            .where(func.lower(Village.name).like(needle))
            .order_by(State.name, District.name, Village.name)
            .limit(limit)
        ).all()
        for v, d, s in villages:
            results.append(
                {
                    "type": "village",
                    "name": v.name,
                    "state": s.name,
                    "district": d.name,
                    "village": v.name,
                    "latitude": v.latitude,
                    "longitude": v.longitude,
                }
            )

        units = db.execute(
            select(AssessmentUnit, District, State)
            .join(District, AssessmentUnit.district_id == District.id)
            .join(State, AssessmentUnit.state_id == State.id)
            .where(func.lower(AssessmentUnit.name).like(needle))
            .order_by(State.name, District.name, AssessmentUnit.name)
            .limit(limit)
        ).all()
        for u, d, s in units:
            results.append(
                {
                    "type": "assessment_unit",
                    "name": u.name,
                    "state": s.name,
                    "district": d.name,
                    "village": None,
                    "latitude": u.latitude,
                    "longitude": u.longitude,
                }
            )

    return {"query": q, "results": results[: limit * 4]}


@router.get("/map")
def gis_map(
    state: str | None = Query(default=None),
    district: str | None = Query(default=None),
    village: str | None = Query(default=None),
    year: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.build_geojson(
        db, state=state, district=district, village=village, year=year, metric=metric
    )


@router.get("/india")
def gis_india(
    year: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.build_india_geojson(db, year=year, metric=metric)


@router.get("/units")
def gis_units(
    state: str | None = Query(default=None),
    district: str | None = Query(default=None),
    village: str | None = Query(default=None),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.unit_centroids(db, state=state, district=district, village=village)


@router.get("/stations")
def gis_stations(
    state: str | None = Query(default=None),
    district: str | None = Query(default=None),
    village: str | None = Query(default=None),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    """Synthetic monitoring-station points (placed district centroids)."""
    return service.unit_centroids(db, state=state, district=district, village=village)


@router.get("/analyze")
def gis_analyze(
    lat: float = Query(...),
    lon: float = Query(...),
    year: int | None = Query(default=None),
    target_year: int | None = Query(default=None),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    """AI analysis for a clicked map location (status, water level, trend,
    risk, prediction and recommendations)."""
    return service.analyze_location(db, lat, lon, year=year, target_year=target_year)


@router.get("/prediction")
def gis_prediction(
    state: str | None = Query(default=None),
    district: str | None = Query(default=None),
    village: str | None = Query(default=None),
    year: int | None = Query(default=None),
    target_year: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.build_prediction_geojson(
        db,
        state=state,
        district=district,
        village=village,
        year=year,
        target_year=target_year,
        metric=metric,
    )


@router.get("/basins")
def gis_basins(
    basin: str | None = Query(default=None),
    year: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    """River-basin choropleth aggregating the districts of each basin."""
    return service.build_basin_geojson(db, basin=basin, year=year, metric=metric)


@router.get("/compare")
def gis_compare(
    state: str | None = Query(default=None),
    district: str | None = Query(default=None),
    village: str | None = Query(default=None),
    year_a: int | None = Query(default=None),
    year_b: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.build_compare(
        db,
        state=state,
        district=district,
        village=village,
        year_a=year_a,
        year_b=year_b,
        metric=metric,
    )


@router.get("/compare/india")
def gis_compare_india(
    year_a: int | None = Query(default=None),
    year_b: int | None = Query(default=None),
    metric: str = Query(default="stage"),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    return service.build_india_compare(db, year_a=year_a, year_b=year_b, metric=metric)


@router.get("/meta")
def gis_meta(
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    meta = service.build_geojson(db)["meta"]
    return {
        "metrics": list(service.METRICS),
        "years": meta["years"],
        "states": meta["states"],
    }

File: frontend/src/services/gis.ts
import { api } from "./api";

export interface MapFeature {
  type: "Feature";
  properties: {
    id: number;
    name: string;
    state: string;
    district: string;
    year: number;
    metric: string;
    metric_value: number | null;
    stage_of_extraction: number | null;
    category: string | null;
    is_demo: boolean;
    latitude: number | null;
    longitude: number | null;
  };
  geometry: {
    type: "Polygon";
    coordinates: number[][][];
  };
}

export interface MapData {
  type: "FeatureCollection";
  features: MapFeature[];
  meta: {
    metric: string;
    year: number;
    years: number[];
    states: string[];
    metrics: string[];
  };
}

export interface CompareProperties {
  id: number;
  name: string;
  state: string;
  district: string;
  year: number;
  metric: string;
  metric_value: number | null;
  metric_value_a: number | null;
  metric_value_b: number | null;
  delta: number | null;
  stage_of_extraction: number | null;
  category: string | null;
  category_a: string | null;
  category_b: string | null;
  category_changed: boolean;
  is_demo: boolean;
  latitude: number | null;
  longitude: number | null;
}

export interface CompareFeature {
  type: "Feature";
  properties: CompareProperties;
  geometry: MapFeature["geometry"];
}

export interface CompareData {
  type: "FeatureCollection";
  features: CompareFeature[];
  meta: {
    metric: string;
    year_a: number;
    year_b: number;
    years: number[];
    states: string[];
    metrics: string[];
  };
}

export interface IndiaCompareFeature {
  type: "Feature";
  properties: Omit<CompareProperties, "id" | "district" | "latitude" | "longitude"> & {
    name: string;
    has_data: boolean;
    unit_count: number;
  };
  geometry: IndiaFeature["geometry"];
}

export interface IndiaCompareData {
  type: "FeatureCollection";
  features: IndiaCompareFeature[];
  meta: CompareData["meta"];
}

export async function getMap(
  params: {
    state?: string;
    district?: string;
    village?: string;
    year?: number;
    metric?: string;
  } = {}
): Promise<MapData> {
  const { data } = await api.get<MapData>("/gis/map", { params });
  return data;
}

export async function getCompare(
  params: {
    state?: string;
    district?: string;
    village?: string;
    year_a?: number;
    year_b?: number;
    metric?: string;
  } = {}
): Promise<CompareData> {
  const { data } = await api.get<CompareData>("/gis/compare", { params });
  return data;
}

export async function getIndiaCompare(
  params: { year_a?: number; year_b?: number; metric?: string } = {}
): Promise<IndiaCompareData> {
  const { data } = await api.get<IndiaCompareData>("/gis/compare/india", { params });
  return data;
}

export interface IndiaFeature {
  type: "Feature";
  properties: {
    name: string;
    has_data: boolean;
    unit_count: number;
    metric_value: number | null;
    stage_of_extraction: number | null;
    category: string | null;
    is_demo: boolean;
    year: number;
    metric: string;
  };
  geometry: {
    type: "MultiPolygon" | "Polygon";
    coordinates: number[][][][] | number[][][];
  };
}

export interface IndiaMapData {
  type: "FeatureCollection";
  features: IndiaFeature[];
  units: MapData;
  meta: MapData["meta"];
}

export async function getIndiaMap(
  params: { year?: number; metric?: string } = {}
): Promise<IndiaMapData> {
  const { data } = await api.get<IndiaMapData>("/gis/india", { params });
  return data;
}

export interface BasinFeature {
  type: "Feature";
  properties: {
    id: number;
    name: string;
    basin: string;
    states: string[];
    unit_count: number;
    district_count: number;
    metric_value: number | null;
    stage_of_extraction: number | null;
    category: string | null;
    is_demo: boolean;
    year: number;
    metric: string;
  };
  geometry: {
    type: "MultiPolygon" | "Polygon";
    coordinates: number[][][][] | number[][][];
  };
}

export interface BasinMapData {
  type: "FeatureCollection";
  features: BasinFeature[];
  meta: MapData["meta"];
}

export async function getBasinMap(
  params: { basin?: string; year?: number; metric?: string } = {}
): Promise<BasinMapData> {
  const { data } = await api.get<BasinMapData>("/gis/basins", { params });
  return data;
}

export interface Station {
  id: number;
  name: string;
  state: string;
  district: string;
  latitude: number;
  longitude: number;
}

export interface LocationAnalysis {
  location: string | null;
  district: string | null;
  state: string | null;
  year: number | null;
  stage: number | null;
  category: string | null;
  water_level: {
    value: number | null;
    unit: string;
    trend_per_year: number | null;
  };
  risk: string | null;
  prediction: {
    target_year: number;
    value: number | null;
    unit: string;
    stage: number | null;
    category: string | null;
  } | null;
  recommendation: string[];
  is_demo: boolean;
  error?: string;
}

export async function getStations(
  params: { state?: string; district?: string; village?: string } = {}
): Promise<Station[]> {
  const { data } = await api.get<Station[]>("/gis/stations", { params });
  return data;
}

export async function getPrediction(
  params: {
    state?: string;
    district?: string;
    village?: string;
    year?: number;
    target_year?: number;
    metric?: string;
  } = {}
): Promise<MapData> {
  const { data } = await api.get<MapData>("/gis/prediction", { params });
  return data;
}

export async function analyzeLocation(
  lat: number,
  lon: number,
  params: { year?: number; target_year?: number } = {}
): Promise<LocationAnalysis> {
  const { data } = await api.get<LocationAnalysis>("/gis/analyze", {
    params: { lat, lon, ...params },
  });
  return data;
}